src/array.rs

```diff
@@ -22,6 +23,142 @@ pub type ReadOnly<T, M> = Generic<T, M, marker::ReadOnly>;
 /// A write-only accessor.
 pub type WriteOnly<T, M> = Generic<T, M, marker::WriteOnly>;
 
+/// Bound wrapper of a single-element accessor.
+/// The lifetime is set to the lifetime of its array accessor.
+pub struct BoundGeneric<'a, T, M, A>
+where
+    M: Mapper,
+    A: AccessorTypeSpecifier + 'static,
+{
+    a: single::Generic<T, Identity, A>,
+    _lifetime: PhantomData<&'a Generic<T, M, A>>,
+}
+
+impl<'a, T, M, A> BoundGeneric<'a, T, M, A>
+where
+    M: Mapper,
+    A: Readable,
+{
+    /// Reads a value from the address that the accessor points to.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if `i >= self.len()`.
+    pub fn read_volatile(&self) -> T {
+        self.a.read_volatile()
+    }
+
+    /// Alias of [`BoundGeneric::read_volatile`].
+    #[deprecated(since = "0.3.1", note = "use `read_volatile`")]
+    pub fn read(&self) -> T {
+        self.read_volatile()
+    }
+}
+impl<'a, T, M, A> BoundGeneric<'a, T, M, A>
+where
+    M: Mapper,
+    A: Writable,
+{
+    /// Writes a value to the address that the accessor points to.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if `i >= self.len()`.
+    pub fn write_volatile(&mut self, v: T) {
+        self.a.write_volatile(v);
+    }
+
+    /// Alias of [`BoundGeneric::write_volatile`].
+    #[deprecated(since = "0.3.1", note = "use `write_volatile`")]
+    pub fn write(&mut self, v: T) {
+        self.write_volatile(v);
+    }
+}
+impl<'a, T, M, A> BoundGeneric<'a, T, M, A>
+where
+    M: Mapper,
+    A: Readable + Writable,
+{
+    /// Updates a value that the accessor points to by reading it, modifying it, and writing it.
+    ///
+    /// Note that some MMIO regions (e.g. the Command Ring Pointer field of the Command
+    /// Ring Control Register of the xHCI) may return 0 regardless of the actual values of the
+    /// fields. For these regions, this operation should be called only once.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if `i >= self.len()`.
+    pub fn update_volatile<U>(&mut self, f: U)
+    where
+        U: FnOnce(&mut T),
+    {
+        self.a.update_volatile(f);
+    }
+
+    /// Alias of [`BoundGeneric::update_volatile`].
+    #[deprecated(since = "0.3.1", note = "use `update_volatile`")]
+    pub fn update<U>(&mut self, f: U)
+    where
+        U: FnOnce(&mut T),
+    {
+        self.update_volatile(f);
+    }
+}
+
+/// A derive macro which converts a field struct type into a struct of accessors with same field names.
+/// For a field struct type `T`, add `#[derive(BoundSetGenericOf)]` before the struct definition to derive
+/// accessor struct type `BoundSetGenericOfT<'a, M, A>`. See [`BoundSetGeneric`] for details.
+pub use accessor_macros::BoundSetGenericOf;
+
+/// Combined with proc-macro [`BoundSetGenericOf`], this trait converts array accessors of field struct types into a struct of accessors with same field names.
```

Comment:
This is an example of
* how can we apply `BoundSetGenericOf` derivation to a struct, and
* how can we spawn a struct of accessors by indexing, out from an array of such struct, invoking `.set_at(i)` method in `accessor::array::BoundSetGeneric` trait.

You can also refer to https://github.com/paulsohn/xhci/commit/a2cf6fbd52bfcf596814e79d64f1f619bf009984 and https://github.com/rust-osdev/xhci/issues/158 (related issue)